propagator/propagator.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagator
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// TraceContextHeaderName is the HTTP header field for Google Cloud Trace
+// https://cloud.google.com/trace/docs/setup#force-trace
+const TraceContextHeaderName = "x-cloud-trace-context"
+
+// TraceContextHeaderFormat is the regular expression pattan for valid Cloud Trace header value
+const TraceContextHeaderFormat = "(?P<trace_id>[0-9a-f]{32})/(?P<span_id>[0-9]{1,20})(;o=(?P<trace_flags>[0-9]))?"
+
+// TraceContextHeaderRe is a regular expression object of TraceContextHeaderFormat.
+var TraceContextHeaderRe = regexp.MustCompile(TraceContextHeaderFormat)
+
+var traceContextHeaders = []string{TraceContextHeaderName}
+
+type CloudTraceFormatPropagator struct{}
+
+func New() propagation.TextMapPropagator {
+	return CloudTraceFormatPropagator{}
+}
+
+func (p CloudTraceFormatPropagator) getHeaderValue(carrier propagation.TextMapCarrier) string {
+	header := carrier.Get(TraceContextHeaderName)
+	if header != "" {
+		return header
+	}
+
+	// Currently, header name can be in different cases (eg. x-cloud-trace-context),
+	// the following part is for handling those cases.
+	for _, key := range carrier.Keys() {
+		if strings.ToLower(key) == TraceContextHeaderName {
+			header = carrier.Get(key)
+			if header != "" {
+				return header
+			}
+		}
+	}
+	return ""
+}
+
+// Inject injects a context to the carrier following Google Cloud Trace format.
+func (p CloudTraceFormatPropagator) Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
+	span := trace.SpanFromContext(ctx)
+	sc := span.SpanContext()
+
+	// https://cloud.google.com/trace/docs/setup#force-trace
+	// Trace ID: 32-char hexadecimal value representing a 128-bit number.
+	// Span ID: decimal representation of the unsigned interger.
+	sidHex := sc.SpanID().String()
+	sid, err := strconv.ParseUint(sidHex, 16, 64)
```

Comment:
This is still parsing the span ID in base 16, though the doc says it's supposed to be decimal.

Reply:
because `sc.SpanID().String()` returns hex value, it needs to be parsed in base 16. this is intended.

Reply:
`sc.SpanID()` is already `[8]byte`, so you could avoid the hex conversion. I suppose it should be BE, so `binary.BigEndian.Uint64(sc.SpanID()[:])` maybe?

Reply:
@punya @aabmass I forgot the use of `encoding/binary`. Yes, you are right. I changed the code to use it.